server: support adding stores to existing nodes

TheSamHuang opened the pull request:
We need to bootstrap additional stores asynchronously. Consider the range that
houses the store ID allocator. When restarting the set of nodes that holds a
quorum of these replicas, when restarting them with additional stores, those
additional stores will require store IDs to get fully bootstrapped. But if
we're gating node start (specifically opening up the RPC floodgates) on having
all stores fully bootstrapped, we'll simply hang when trying to allocate store
IDs.  See TestAddNewStoresToExistingNodes and #39415 for more details.

Instead we opt to bootstrap additional stores asynchronously, and rely on the
blocking function to signal to the caller that all stores have been fully
bootstrapped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55350)
<!-- Reviewable:end -->

tbg reviewed (commented):
Looks good! Does it hang as we expect?

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-:-MJBNerq9qPYVVzxnHev:b-rl55ap)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @TheSamHuang)

___
*[pkg/server/node_test.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJBNHy_Bk_jym8Eh_2c:-MJBNHya5UVzcbsX2YG7:b-qemtjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a43323cce128f81eb06282c39f585dff437fe16/pkg/server/node_test.go#L171)):*
> ```Go
> // TestAddNewStoresToExistingNodes starts a cluster with three nodes,
> // shuts down all nodes and adds a store to each node, and ensures
> // nodes start back up successfully
> ```

nit: `.`
___
*[pkg/server/node_test.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJBNMVA6Ac18MICMIyt:-MJBNMVA6Ac18MICMIyu:b3kqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a43323cce128f81eb06282c39f585dff437fe16/pkg/server/node_test.go#L204)):*
> ```Go
> 	tc.Stopper().Stop(ctx)
> 
> 	// Add an additional store to each node
> ```

.
___
*[pkg/server/node_test.go, line 235 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJBNQYmDfS9mUeHbcpr:-MJBNQYmDfS9mUeHbcps:b3kqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a43323cce128f81eb06282c39f585dff437fe16/pkg/server/node_test.go#L235)):*
> ```Go
> 	}
> 
> 	// Start all nodes with additional stores
> ```

.
___
*[pkg/server/node_test.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJBNSt2A-8n2yAuNNFF:-MJBNSt2A-8n2yAuNNFG:by2x5nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a43323cce128f81eb06282c39f585dff437fe16/pkg/server/node_test.go#L241)):*
> ```Go
> 	// Ensure all nodes have 2 stores available
> 	testutils.SucceedsSoon(t, func() error {
> 		for _, idx := range []int{0, 1, 2} {
> ```

Might as well `for _, srv := range tc.Servers { ... }`


<!-- Sent from Reviewable.io -->

TheSamHuang commented:
> Looks good! Does it hang as we expect?

Yes, it does!

TheSamHuang commented:
After trying to debug why the new test was failing (in hopes of solving the stores issue), I found out that the testing libraries do not allow for starting up a cluster or even an individual server after they have been stopped. (After calling `testCluster.Stopper().Stop()` or `server.Stopper().Stop()`, `testCluster.StartTestCluster()` and `server.Start()` will always fail). Since this unit test will always fail, I will opt for writing a roachtest instead.

tbg commented:
I think something got lost in translation when we talked about this test. I did mention that you can't restart an existing server, but that's why we're using the disks - we will start a new server from the existing stores. That *should* work and if it doesn't, I'm happy to help fix it. I don't think we should throw the towel yet and write a roachtest here.

tbg commented:
Honestly looking at your code I'm a bit confused by your comment. You're already doing the right thing in that test, which intentionally doesn't work (since you haven't fixed the store creation issue yet). I was trying out this branch and there are some issues to chew through locally. For one, for some reason once you give disks initial cluster start takes 10+ seconds - I'm sure I looked at this before and couldn't figure it out, so now is the time to double down - and you need the `ParallelStart: true` setting when restarting the cluster.

I'll put together a diff for you when I've gotten somewhere.

tbg commented:
@TheSamHuang I updated your branch with commits that "fix" both the issues with the test and make the change to bootstrap additional stores asynchronously. Interestingly, this is still failing, but with an unrelated storage issue:

```
anic: open /tmp/TestAddNewStoresToExistingNodes119630251/000304.log: no such file or directory [recovered]
	panic: open /tmp/TestAddNewStoresToExistingNodes119630251/000304.log: no such file or directory

goroutine 5848 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00316a000, 0x5933c40, 0xc0031d15f0)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:207 +0x11f
panic(0x4442840, 0xc001cfcff0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/pebble.(*DB).makeRoomForWrite(0xc000a7d800, 0xc001252140, 0x14426, 0x16000)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:1393 +0x119d
github.com/cockroachdb/pebble.(*DB).commitWrite(0xc000a7d800, 0xc001252140, 0x0, 0x0, 0xc00384e3a8, 0xc001252140, 0xc00384e3a8)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:628 +0xbb
github.com/cockroachdb/pebble.(*commitPipeline).prepare(0xc00384e000, 0xc001252140, 0x0, 0x5933c40, 0xc003f9bc50, 0x5a8e640)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/commit.go:377 +0x162
github.com/cockroachdb/pebble.(*commitPipeline).Commit(0xc00384e000, 0xc001252140, 0x0, 0x0, 0x20)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/commit.go:253 +0x73
github.com/cockroachdb/pebble.(*DB).Apply(0xc000a7d800, 0xc001252140, 0x80adb00, 0xc002d42a00, 0xc00026b800)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:556 +0x10a
github.com/cockroachdb/pebble.(*Batch).Commit(...)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/batch.go:727
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleBatch).Commit(0xc0018bec00, 0x5933c00, 0xc003f9bc50, 0x0)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_batch.go:341 +0x5e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).ApplyToStateMachine(0xc00026b8c0, 0x5933c40, 0xc003f9bc50, 0x7f4d574be180, 0xc00026bad0)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_state_machine.go:868 +0xce
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).applyOneBatch(0xc00205d798, 0x5933c40, 0xc003f9bc50, 0x5976080, 0xc00026ba70, 0x0, 0x0)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:286 +0x1d7
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).ApplyCommittedEntries(0xc00205d798, 0x5933c40, 0xc003f9bc50, 0x2, 0x49a7643)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:247 +0xb9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(0xc00026b800, 0x5933c40, 0xc003f9bc50, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:765 +0xff9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(0xc00026b800, 0x5933c40, 0xc003f9bc50, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
```

This occurs in the test when we restart the cluster from the store directories we made. Uh oh. @petermattis I hope this is something silly related to the test setup. Can you have the storage team take a look?

There's also a second failure mode:

>     testcluster.go:260: failed to register engines with debug server: store has not been bootstrapped

This is because https://github.com/cockroachdb/cockroach/blob/7bed111566b5657e7c2d44cab792b588f9ae9973/pkg/server/debug/server.go#L221-L230 expects all stores bootstrapped by the time we register them, but this is not true any more. I'm adding a hacky commit to avoid this failure mode for now.

tbg commented:
Actually, the pebble issue doesn't show up any more, despite a few minutes of `make stress PKG=./pkg/server/ TESTS=TestAddNewStoresToExisting STRESSFLAGS='-p 4'`. Unclear what's going on here. Maybe build jank? I was running the test through Goland which isn't hooked up to the makefile.

tbg commented:
Interesting. Every time? I got it once and then never again.

irfansharif commented:
Nope, not every time. I get the other debug server error too, and manually ignoring that error it's hard to repro, but I haven't really tried in earnest.

TheSamHuang commented:
> I think something got lost in translation when we talked about this test. I did mention that you can't restart an existing server, but that's why we're using the disks - we will start a new server from the existing stores. That _should_ work and if it doesn't, I'm happy to help fix it. I don't think we should throw the towel yet and write a roachtest here.

Thanks, Tobias. It slipped my mind that we should start a new server from existing stores instead of restarting an existing one -- sorry about that!

irfansharif reviewed (changes requested):
Thanks for working on this thus far Sam! I think we're super close, I've included in comments below the few changes we'll need to make here before wrapping it all up. I'm excited to have crdb (once again, ha) attain the ability to support nodes with multiple stores. 

We'll also want to clean up our commit history here. We'll want to separate out @tbg's changes to the testing infrastructure as the initial commits, and then introduce the test (nicely done by the way) + the async bootstrap process as a single final commit. Lastly, we'll want to change the commit title to be "server: support adding stores to existing nodes". Our convention for the prefix is to have it match the package name in question (for us it's pkg/server).

irfansharif commented on pkg/server/connectivity_test.go:
```diff
@@ -390,3 +391,115 @@ func TestDecommissionedNodeCannotConnect(t *testing.T) {
+// multiple stores per node, in particular when new stores are
+// added while nodes are shut down. This test starts a cluster with
+// three nodes, shuts down all nodes and adds a store to each node,
+// and ensures nodes start back up successfully. See:
```
"See #39415".

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +451,32 @@ func (n *Node) start(
-			return err
-		}
+		// We need to bootstrap additional stores asynchronously because we can't rely on
+		// the KV layer to be available yet. See:
```
Ditto.

irfansharif commented on pkg/server/server.go:
```diff
@@ -1603,8 +1604,15 @@ func (s *Server) PreStart(ctx context.Context) error {
+	// Block until all stores are fully bootstrapped.
+	log.Infof(ctx, "waiting for all stores to be fully bootstrapped")
+	waitForBootstrapNewStores()
+	log.Infof(ctx, "all stores fully bootstrapped")
```
Let's remove this logging statement, it's not necessary given we're logging right after so can infer that everything has been fully bootstrapped.

irfansharif commented on pkg/server/node.go:
```diff
@@ -348,7 +348,7 @@ func (n *Node) start(
 	localityAddress []roachpb.LocalityAddress,
 	nodeDescriptorCallback func(descriptor roachpb.NodeDescriptor),
-) error {
+) (error, func()) {
```
Let's use named return values here: https://tour.golang.org/basics/7. Let's also add some commentary above explaining what this new return value is intended for.

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +451,32 @@ func (n *Node) start(
-	// TODO(tbg): address https://github.com/cockroachdb/cockroach/issues/39415.
-	// Should be easy enough. Writing the test is probably most of the work.
+	// Bootstrap any uninitialized stores and signal channel once complete.
+	bootstrapStoresComplete := make(chan struct{})
```
s/bootstrapStoresComplete/bootstrapNewStoresDone

irfansharif replied:
[nit] I think what we could do instead here is something like the following:
```go
var waitForBootstrapNewStores := func() { }
if len(state.newEngines) > 0 {
  bootstrapStoresComplete := make(chan struct{})
  waitForBootstrapNewStores = func() { <-bootstrapStoresComplete }
  // ...
}

// ...
return nil, waitForBootstrapNewStores
```

What do ya think? I think it localizes usage of the signal channel to just the conditional where it's needed.

TheSamHuang replied:
I think that's definitely better! I'll make the change.

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +451,32 @@ func (n *Node) start(
+			}
+			close(bootstrapStoresComplete)
+		})
+	} else {
```
Add a comment here just spelling out that we're "done" seeing as how there aren't any new stores to be added.

irfansharif commented on pkg/server/server.go:
```diff
@@ -1603,8 +1604,15 @@ func (s *Server) PreStart(ctx context.Context) error {
 	s.grpc.setMode(modeOperational)
 
+	// Block until all stores are fully bootstrapped.
+	log.Infof(ctx, "waiting for all stores to be fully bootstrapped")
```
I'd actually remove this logging statement as well. It's a bit confusing if you consider that we'll also be logging this when restarting nodes with already bootstrapped stores (from the very first start). So, why are we waiting for anything?

irfansharif reviewed (commented):
Nice, almost there. We just have our git commit history to unwrangle and we should be good to go. Happy to help you with it offline if needed!

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +453,33 @@ func (n *Node) start(
+		// Create bootstrapNewStoresDone channel and update waitForBootstrapNewStores function to block on it.
+		bootstrapNewStoresDone := make(chan struct{})
+		waitForBootstrapNewStores = func() { <-bootstrapNewStoresDone }
+
```
Remove this stray newline.

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +453,33 @@ func (n *Node) start(
+		// We need to bootstrap additional stores asynchronously because we can't rely on
+		// the KV layer to be available yet. See #39415.
+
+		// Create bootstrapNewStoresDone channel and update waitForBootstrapNewStores function to block on it.
```
[nit] The comment here is far too identical to the code right below, so doesn't necessarily add much to it. Given our variable names (they're also documentation in some sense), I think it's fine to skip it entirely.

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +453,33 @@ func (n *Node) start(
-	//
-	// TODO(tbg): address https://github.com/cockroachdb/cockroach/issues/39415.
-	// Should be easy enough. Writing the test is probably most of the work.
+	// Bootstrap any uninitialized stores and signal channel once complete.
```
[nit] Let's remove the "and signal channel once complete" portion here given we're defining the channel elsewhere. We can add here that "this function remains a no-op unless we find ourselves bootstrapping new stores".

irfansharif commented:
Bit of update here. Upon stressing the new test it looks like even Sam is running into opaque storage errors. This time it was a diskstall. We're going to clean up the history here tomorrow and then punt the test off to #storage to see if it's obvious to them why what we're doing here is problematic. Nothing immediately jumps out to me. I suspect we're tickling some test harness bug in trying to use multiple stores, but that's just a guess.

```
F201019 19:35:52.330137 6656 storage/pebble.go:597  [n3] disk stall detected: pebble unable to write to /var/folders/4c/vf8t7zk96g192f8hx24_hgm40000gn/T/TestAddNewStoresToExistingNodes907524864/000183.log in 2139.66 seconds
goroutine 6656 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0xb8f6600, 0xc005dff900, 0x40e8902, 0xc002a9c5b0)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/util/log/get_stacks.go:25 +0xb8
github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(0xb8f5dc0, 0x4, 0x163f7bd636396da8, 0x1a00, 0xb4b7c12, 0x11, 0x255, 0xc0001684d0, 0xa6, 0xc000231d80, ...)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:241 +0xa02
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x9290b60, 0xc000d23cb0, 0x4, 0x2, 0x836c724, 0x41, 0xc001158638, 0x2, 0x2)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:52 +0x1ac
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x9290b60, 0xc000d23cb0, 0x1, 0xc000000004, 0x836c724, 0x41, 0xc001158638, 0x2, 0x2)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:44 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(...)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:165
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).connectEventMetrics.func1(0xc000030000, 0x64, 0x1f22d6da9b8)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:597 +0x2c8
github.com/cockroachdb/pebble.(*Options).EnsureDefaults.func1(0xc000030000, 0x64, 0x1f22d6da9b8)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/options.go:562 +0x47
github.com/cockroachdb/pebble/vfs.diskHealthCheckingFS.ReuseForWrite.func1(0x1f22d6da9b8)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:173 +0x41
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1(0xc004368bc0)
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:77 +0x2cd
created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker
	/Users/samhuang/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:58 +0x50```

irfansharif committed:
server: support adding stores to existing nodes

We need to bootstrap additional stores asynchronously. Consider the range that
houses the store ID allocator. When restarting the set of nodes that holds a
quorum of these replicas, when restarting them with additional stores, those
additional stores will require store IDs to get fully bootstrapped. But if
we're gating node start (specifically opening up the RPC floodgates) on having
all stores fully bootstrapped, we'll simply hang when trying to allocate store
IDs.  See TestAddNewStoresToExistingNodes and #39415 for more details.

Instead we opt to bootstrap additional stores asynchronously, and rely on the
blocking function to signal to the caller that all stores have been fully
bootstrapped.

Release note: None

Co-authored-by: Sam Huang <samh@cockroachlabs.com>
Co-authored-by: Tobias Grieger <tobias.b.grieger@gmail.com>

TheSamHuang commented:
`TestAddNewStoresToExistingNodes` should not be run under stress so I added `skip.UnderStress(t, "too many new stores and nodes for stress")` (running it under stress leads to a disk stall after a large number of runs).

In addition, moved `TestAddNewStoresToExistingNodes` to its own file as none of the existing test files seemed appropriate for it.

irfansharif commented on pkg/server/node.go:
```diff
@@ -451,25 +453,32 @@ func (n *Node) start(
-			return err
-		}
+		// We need to bootstrap additional stores asynchronously because we
+		// can't rely on the KV layer to be available yet. See #39415.
```
Let's use more words here. I like what we have in the commit message, so amending that, how about the following?

```
We need to bootstrap additional stores asynchronously. Consider the range that
houses the store ID allocator. When restarting the set of nodes that holds a
quorum of these replicas, when restarting them with additional stores, those
additional stores will require store IDs to get fully bootstrapped. But if we're
gating node start (specifically opening up the RPC floodgates) on having all
stores fully bootstrapped, we'll simply hang when trying to allocate store IDs.
See TestAddNewStoresToExistingNodes and #39415 for more details.

Instead we opt to bootstrap additional stores asynchronously, and rely on the
blocking function to signal to the caller that all stores have been fully
bootstrapped.
```

Feel free to include the same in the commit message as well.

irfansharif commented on pkg/server/server.go:
```diff
@@ -1603,8 +1604,13 @@ func (s *Server) PreStart(ctx context.Context) error {
+	// bootstrapped.
 	s.grpc.setMode(modeOperational)
 
+	// Block until all stores are fully bootstrapped.
```
```
We'll block here until all stores are fully bootstrapped. We do this here for
two reasons:
- some of the components below depend on all stores being fully bootstrapped
  (like the debug server registration for e.g.)
- we'll need to do it after having opened up the RPC floodgates (due to the
  hazard described in Node.start, around bootstrapping additional stores)
```

irfansharif reviewed (approved):
LGTM! Let's get a final set of eyes from @tbg before moving on from here. Thanks for all the work Sam.

tbg reviewed (commented):
+1 on Irfan's comments about comments. This bootstrap problem is not well known outside of the reviewers of this PR. It should be clear as day to anyone visiting this area in the future.

> Reviewed 4 of 6 files at r4, 2 of 5 files at r11, 3 of 3 files at r16, 3 of 3 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-:-MK9E8nh2dYfRZ1Ks3Ps:b-1jce0a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @tbg, and @TheSamHuang)

___
*[pkg/server/node.go, line 351 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJbMaYgLU4wjRlH0JBj-r14-351:-MK9CZ890MFLGHmBD-Cn:bi6gkxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7f4afdb2bef2de267e1062ce301e72bb7dbe726/pkg/server/node.go#L351)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Let's use named return values here: https://tour.golang.org/basics/7. Let's also add some commentary above explaining what this new return value is intended for.
</blockquote></details>

By convention, the error always goes last. Also, it doesn't need to be named:

`(waitForBootstrapNewStores func(), _ error)`

I'm also confused why we stopped returning the channel. A blocking function is IMO a worse choice than a channel. (You've probably discussed this with Irfan before). The upside of a channel is that it provides better cancel-ability for the waiter.

I think there's also a third option that seems appealing, which is that instead of returning something from `(*Node).start`, we add a new method on `*Node` that does the waiting:

```go
// bootstrapNewStoresCh returns a channel that is closed once all
// additional empty stores have been assigned a StoreIdent.
func (n *Node) bootstrapNewStoresCh() <-chan struct{} {
   return n.bootstrapNewStoresCh
}
```
___
*[pkg/server/node.go, line 476 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MK9C8DZ6rY1RZtfhOLw:-MK9C8DZ6rY1RZtfhOLx:barzpfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6256f4f03bc09118c0d9f93ab7aca27292a0e7b/pkg/server/node.go#L476)):*
> ```Go
> 		bootstrapNewStoresDone := make(chan struct{})
> 		waitForBootstrapNewStores = func() { <-bootstrapNewStoresDone }
> 		_ = n.stopper.RunAsyncTask(ctx, "bootstrap-stores", func(ctx context.Context) {
> ```

if RunAsyncTask returns an error, we need (or should) close the channel as well.
___
*[pkg/server/server.go, line 1543 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55350#-MJbpKf2Yg08nFVqEscj-r11-1543:-MK9DwCw0dB4_ABN9lbz:bew7xk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca7f557f7ed569f58dab36684ec27aab8722804/pkg/server/server.go#L1543)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Instead of constructing the channel at the caller, we can do it within `node.start` itself. What we can do instead is return a function like `waitForBootstrapNewStores` and call that instead, and within that function body (unbeknownst to the caller) it would wait for the right channel.

As for why we should structure it this way: it lets us contain the scope of this signal channel to just the bits we care about. This server.PreStart method is already too long (as you're painfully aware), so the fewer things in scope, the clearer things are.
</blockquote></details>

Ah, this is where the suggestion to return a closure comes from. Curious why you prefer this over a `<-chan struct{}` which imo is the most functional representation of "a thing that signals you".


<!-- Sent from Reviewable.io -->

TheSamHuang commented:
Thanks Toby and Irfan for the comments! 

> I think there's also a third option that seems appealing, which is that instead of returning something from (*Node).start, we add a new method on *Node that does the waiting
`// bootstrapNewStoresCh returns a channel that is closed once all
// additional empty stores have been assigned a StoreIdent.
func (n *Node) bootstrapNewStoresCh() <-chan struct{} {
   return n.bootstrapNewStoresCh
}`

I made one change to your suggestion - instead of defining a function that simply returns the `bootstrapNewStoresCh`, I defined a function on node that blocks until the stores are bootstrapped. This way, it's a bit more clear to the caller what is happening: they can simply call `s.node.waitForBootstrapNewStores()`. What do you think? 


> if RunAsyncTask returns an error, we need (or should) close the channel as well. 

Instead of closing the channel, I instead opted to `log.Fatalf()` instead because if the async task to bootstrap the new stores failed, we should fail here rather than fail transparently to the caller. If the async task failed, I think the node would fail later on by getting stuck in a loop like we were seeing. It would be harder to debug/troubleshoot in that case, than just logging and crashing when the async task failed.

irfansharif commented:
>  Curious why you prefer this over a <-chan struct{} which imo is the most functional representation of "a thing that signals you".

/shrug no good reason other than generally wanting to avoid channels in APIs.

> s.node.waitForBootstrapNewStores()

I'm fine either way, I didn't have strong opinions on what we had before either.

> Instead of closing the channel, I instead opted to log.Fatalf() instead because if the async task to bootstrap the new stores failed, we should fail here rather than fail transparently to the caller. 

I believe Tobi was talking about the error returned here, which we're currently ignoring:
```go
_ = n.stopper.RunAsyncTask
```

This isn't from bootstrapping the stores themselves, it's from not being able to start off the async task in the first place. We'll need to handle it appropriately.

TheSamHuang commented:
>This isn't from bootstrapping the stores themselves, it's from not being able to start off the async task in the first place. We'll need to handle it appropriately.

Yes, I agree. I think it will be more clear what I meant when I push the amended commit. I meant that if starting the async task failed, the stores wouldn't get bootstrapped so we should fail.

TheSamHuang commented:
> > This isn't from bootstrapping the stores themselves, it's from not being able to start off the async task in the first place. We'll need to handle it appropriately.
> 
> Yes, I agree. I think it will be more clear what I meant when I push the amended commit. I meant that if starting the async task failed, the stores wouldn't get bootstrapped so we should fail.

After some more thought, decided to go with closing the channel and just returning the error to the caller instead!

irfansharif commented on pkg/server/node.go:
```diff
@@ -333,11 +334,13 @@ func (n *Node) AnnotateCtxWithSpan(
+// Launches periodic store gossiping in a goroutine. A callback can
+// be optionally provided that will be invoked once this node's
+// NodeDescriptor is available, to help bootstrapping. A
+// waitForBootstrapNewStores function is returned to allow the caller to
```
This last sentence is no longer applicable.

irfansharif commented on pkg/server/node.go:
```diff
@@ -492,6 +512,14 @@ func (n *Node) start(
 	return nil
 }
 
+// waitForBootstrapNewStores waits on a channel that is closed once all
```
"waitForBootstrapNewStores blocks until all additional stores, if any, have been bootstrapped."

irfansharif reviewed (approved):
LGTM mod comment adjustments below.

TheSamHuang commented:
bors r=irfansharif
